File: LeetCode/5.Longest_Palindromic_Substring/5.Longest_Palindromic_Substring.py
#!/usr/bin/env python3


# dp
class Solution:
    def longestPalindrome(self, s: str) -> str:
        # l,r  r >= l
        # l == r dp[l][r] = True
        # l == r-1 if s[l] == s[r] : dp[l][r] == True
        # dp[l][r] = (s[l] == s[r]) and dp[l+1][r-1]

        n = len(s)
        start, max_len = 0, 1

        # init dp
        dp = [[False] * n for _ in range(n)]

        # 'a' every character itself is pali / two characters
        for i in range(n):
            dp[i][i] = True
            if i + 1 < n and s[i] == s[i + 1]:
                dp[i][i + 1] = True
                if max_len < 2:
                    start, max_len = i, 2

        # iterate  l + 1 < r
        for length in range(3, n + 1):
            for l in range(n - length + 1):
                r = l + length - 1
                print(f"l:{l}, r:{r}")
                if s[l] == s[r] and dp[l + 1][r - 1]:
                    dp[l][r] = True
                    if length > max_len:
                        start, max_len = l, length

        return s[start : start + max_len]


# expand around center
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)

        def expand(l, r):
            while 0 <= l < n and 0 <= r < n and s[l] == s[r]:
                l -= 1
                r += 1

            return l + 1, r - 1

        start, max_len = 0, 1

        for i in range(n):
            # odd
            l1, r1 = expand(i, i)
            if r1 - l1 + 1 > max_len:
                start, max_len = l1, r1 - l1 + 1

            # even
            if i + 1 < n:
                l2, r2 = expand(i, i + 1)
                if r2 - l2 + 1 > max_len:
                    start, max_len = l2, r2 - l2 + 1

        return s[start : start + max_len]

File: LeetCode/5.Longest_Palindromic_Substring/5.Longest_Palindromic_Substring.swift
class Solution {
    func longestPalindrome(_ s: String) -> String {
        let chs = Array(s)
        let n = chs.count

        var (start, len) = (0, 1)

        func helper(_ l: Int, _ r: Int) {
            var l = l, r = r

            while l >= 0 && r < n && chs[l] == chs[r] {
                let curLen = r - l + 1
                if curLen > len {
                    (start, len) = (l, curLen)
                }
                l -= 1
                r += 1
            }
        }

        for i in 0 ..< n {
            helper(i, i)
            if i > 0 {
                helper(i - 1, i)
            }
        }

        return String(chs[start ..< start + len])
    }
}

// DP
class Solution {
    func longestPalindrome(_ s: String) -> String {
        let chs = Array(s)
        let n = chs.count

        var dp = Array(repeating: Array(repeating: false, count: n), count: n)

        var (start, len) = (0, 1)

        // init dp
        for i in 0 ..< n {
            dp[i][i] = true
            if i > 0 && chs[i - 1] == chs[i] {
                dp[i - 1][i] = true
                if len < 2 {
                    (start, len) = (i - 1, 2)
                }
            }
        }

        if n >= 3 {
            for l in 3 ... n {
                for left in 0 ..< n {
                    let right = left + l - 1
                    if right >= n { break }

                    dp[left][right] = dp[left + 1][right - 1] && (chs[left] == chs[right])
                    if dp[left][right] && l > len {
                        (start, len) = (left, l)
                    }
                }
            }
        }

        return String(chs[start ..< start + len])
    }
}
